### PR TITLE
Migrate Details component to TSX

### PR DESCRIPTION
**Pull request by VanAnderson**
This PR migrates the `Details` component and `useDetails` hook to TypeScript as part of the [TypeScript refactor](https://github.com/primer/components/issues/970).

### CODE REVIEW

File: src/hooks/useDetails.tsx [hunk 1]
```diff
@@ -1,13 +1,23 @@
 import {useCallback, useEffect, useState, useRef} from 'react'
 
-function useDetails({ref, closeOnOutsideClick, defaultOpen, onClickOutside} = {}) {
+type useDetailsParameters = {
+  ref?: {current: HTMLElement}
```
Comment (VanAnderson) [inline src/hooks/useDetails.tsx:4]: I'm wondering if there is an existing type for refs that I can import or if it's just best to do it inline like I have it here?

Comment (colebemis) [reply src/hooks/useDetails.tsx:4]: This should do the trick :)

```suggestion
  ref?: React.RefObject<HTMLElement>
```

File: src/hooks/useDetails.tsx [hunk 2]
```diff
@@ -1,13 +1,23 @@
 import {useCallback, useEffect, useState, useRef} from 'react'
 
-function useDetails({ref, closeOnOutsideClick, defaultOpen, onClickOutside} = {}) {
+type useDetailsParameters = {
```
Comment (colebemis) [inline src/hooks/useDetails.tsx:3]: I think it's a convention to capitalize the first letter of types:

```suggestion
type UseDetailsParameters = {
```

File: src/hooks/useDetails.tsx [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+  ref?: {current: HTMLElement}
+  closeOnOutsideClick?: boolean
+  defaultOpen?: boolean
+  onClickOutside?: Function
```
Comment (colebemis) [inline src/hooks/useDetails.tsx:7]: I think we can be more specific here:

```suggestion
  onClickOutside?: (event: MouseEvent) => void
```

File: src/__tests__/Details.tsx [hunk 4]
```diff
@@ -92,7 +93,7 @@ describe('Details', () => {
 
   it('Does not toggle when you click inside', () => {
     const Component = () => {
-      const {getDetailsProps} = useDetails({closeOnOutsideClick: true, defaultOpen: true})
+      const {getDetailsProps, open} = useDetails({closeOnOutsideClick: true, defaultOpen: true})
       return (
         <Details {...getDetailsProps}>
```
Comment (colebemis) [inline src/__tests__/Details.tsx:98]: Interesting. It looks like none of the Details tests are using the `getDetailsProps` function correctly. They all need to updated to call the function and pass the return value to the Details component:

```suggestion
        <Details {...getDetailsProps()}>
```

Comment (colebemis) [reply src/__tests__/Details.tsx:98]: You'll also probably need to update the prop type of Details to accept the return value of `getDetailsProps`

Comment (VanAnderson) [reply src/__tests__/Details.tsx]: Tried to accomplish this in https://github.com/primer/components/pull/1026/commits/57dafdda03b40868dc1976494713d506a864a096 and I'm getting a TS error. About to sign off here, but I might sync up with you in the morning to see if we can take a look at that.

**Review by colebemis — APPROVED** [review]

**Pull request merged**